mmdet/datasets/pipelines/transforms.py

```diff
@@ -1903,3 +1905,195 @@ def __repr__(self):
+            will be removed. Default: 2.
+        area_ratio_filter_thr (tuple): Threshold of area ratio between
+            original bboxes and wrapped bboxes. If smaller than this value,
+            the box will be removed. Default: 0.2.
+        aspect_ratio_filter_thr (int): Aspect ratio of width and height
+            threshold to filter bboxes. If max(h/w, w/h) larger than this
+            value, the box will be removed.
+    """
+
+    def __init__(self,
+                 rotate_degrees=10.0,
+                 translate_ratio=0.1,
+                 scaling_ratio=(0.5, 1.5),
+                 shear_degrees=2.0,
+                 border=(0, 0),
+                 border_val=(114, 114, 114),
+                 wh_filter_thr=2,
+                 area_ratio_filter_thr=0.2,
+                 aspect_ratio_filter_thr=20):
+        assert 0 <= translate_ratio <= 1
+        assert scaling_ratio[0] <= scaling_ratio[1]
+        assert scaling_ratio[0] > 0
+        self.rotate_degrees = rotate_degrees
+        self.translate_ratio = translate_ratio
+        self.scaling_ratio = scaling_ratio
+        self.shear_degrees = shear_degrees
+        self.border = border
+        self.border_val = border_val
+        self.wh_filter_thr = wh_filter_thr
+        self.area_ratio_filter_thr = area_ratio_filter_thr
+        self.aspect_ratio_filter_thr = aspect_ratio_filter_thr
+
+    def __call__(self, results):
+        img = results['img']
+        height = img.shape[0] + self.border[0] * 2
+        width = img.shape[1] + self.border[1] * 2
+
+        # Center
+        center_matrix = np.eye(3)
+        center_matrix[0, 2] = -img.shape[1] / 2  # x translation (pixels)
+        center_matrix[1, 2] = -img.shape[0] / 2  # y translation (pixels)
+
+        # Rotation
+        rotation_degree = random.uniform(-self.rotate_degrees,
+                                         self.rotate_degrees)
+        rotation_matrix = self._get_rotation_matrix(rotation_degree)
+
+        # Scaling
+        scaling_ratio = random.uniform(self.scaling_ratio[0],
+                                       self.scaling_ratio[1])
+        scaling_matrix = self._get_scaling_matrix(scaling_ratio)
+
+        # Shear
+        x_degree = random.uniform(-self.shear_degrees, self.shear_degrees)
+        y_degree = random.uniform(-self.shear_degrees, self.shear_degrees)
+        shear_matrix = self._get_shear_matrix(x_degree, y_degree)
+
+        # Translation
+        trans_x = random.uniform(0.5 - self.translate_ratio,
+                                 0.5 + self.translate_ratio) * width
+        trans_y = random.uniform(0.5 - self.translate_ratio,
+                                 0.5 + self.translate_ratio) * height
+        translate_matrix = self._get_translation_matrix(trans_x, trans_y)
+
+        warp_matrix = (
+            translate_matrix @ shear_matrix @ rotation_matrix @ scaling_matrix
+            @ center_matrix)
+
+        img = cv2.warpPerspective(
```

Comment:
This is Perspective, not affine. Maybe we should support at the same time.

Reply:
When the third row of the homography matrix is [0,0,1], warpPerspective is equivalent to warpAffine mathematically.

Reply:
But `warpPerspective` and `RandomAffine` are inconsistent.